File: 2022/day_3/task_1.pl
#!/usr/bin/perl

use experimental 'smartmatch';

sub uniq_vals {
    my %seen;
    grep !$seen{$_}++, @_;
}

open(file_data, "<input.txt") or die "Couldn't open file: $!";

@a_z = ("a".."z");
$total = 0;

while(<file_data>) {

   $_ =~ s/^\s+|\s+$//g;
   $h = int(length($_)/2);

   @l = split //, $_;

   @f = @l[0..$h-1];
   @s = @l[$h..length($_)];

   @r = uniq_vals(grep { $f = $_; grep $_ eq $f, @s } @f);

   foreach $e (@r){

       ($index) = grep { $a_z[lc $_] ~~ lc $e } 0 .. $#a_z;
        $total += ($e eq uc $e)? 27 + $index : 1 + $index;
   }
}
print "Sum: $total\n";

File: 2023/day_4/task_1.py
to_int = lambda x: set([int(i) for i in x.split(' ') if i])
points = lambda values: 0 if len(values) == 0 else 1*2**(len(values)-1)

cards = list([l.strip().split(':')[1:][0] for l in open('input.txt')])

total = 0
for card in [c.split('|') for c in cards]:
    total += points(to_int(card[0]) & to_int(card[1]))
print("Result:", total)
